packages/react-urql/src/hooks/useFragment.ts

```diff
@@ -0,0 +1,345 @@
+   */
+  fetching: boolean;
+  /** The data for the masked fragment. */
+  data?: Data | null;
+}
+
+const isSuspense = (client: Client, context?: Partial<OperationContext>) =>
+  context && context.suspense !== undefined
+    ? !!context.suspense
+    : client.suspense;
+
+/** Hook to mask a GraphQL Fragment given its data. (BETA)
+ *
+ * @param args - a {@link UseFragmentArgs} object, to pass a `fragment` and `data`.
+ * @returns a {@link UseFragmentState} result.
+ *
+ * @remarks
+ * `useFragments` allows GraphQL fragments to mask their data.
+ * Given {@link UseFragmentArgs.query} and {@link UseFragmentArgs.data}, it will
+ * return the masked data for the fragment contained in query.
+ *
+ * Additionally, if the `suspense` option is enabled on the `Client`,
+ * the `useFragment` hook will suspend instead of indicating that it’s
+ * waiting for a result via {@link UseFragmentState.fetching}.
+ *
+ * @example
+ * ```ts
+ * import { gql, useFragment } from 'urql';
+ *
+ * const TodoFields = gql`
+ *   fragment TodoFields on Todo { id name }
+ * `;
+ *
+ * const Todo = (props) => {
+ *   const result = useFragment({
+ *     data: props.todo,
+ *     query: TodoFields,
+ *   });
+ *   // ...
+ * };
+ * ```
+ */
+export function useFragment<Data>(
+  args: UseFragmentArgs<Data>
+): UseFragmentState<Data> {
+  const client = useClient();
+  const cache = getFragmentCacheForClient(client);
+  const suspense = isSuspense(client, args.context);
+  const request = useRequest(args.query, args.data || {});
+
+  const fragments = React.useMemo(() => {
+    return request.query.definitions.reduce<
+      Record<string, FragmentDefinitionNode>
+    >((acc, frag) => {
+      if (frag.kind === Kind.FRAGMENT_DEFINITION) {
+        acc[frag.name.value] = frag;
+      }
+      return acc;
+    }, {});
+  }, [request.query]);
+
+  const fragment = React.useMemo(() => {
+    return request.query.definitions.find(
+      x =>
+        x.kind === Kind.FRAGMENT_DEFINITION &&
+        ((args.name && x.name.value === args.name) || !args.name)
+    ) as FragmentDefinitionNode | undefined;
+  }, [request.query, args.name]);
+
+  if (!fragment) {
+    throw new Error(
+      'Passed document did not contain a fragment definition' + args.name
+        ? ` for ${args.name}.`
+        : '.'
+    );
+  }
+
+  const getSnapshot = React.useCallback(
+    (
+      request: GraphQLRequest<Data, AnyVariables>,
+      data: Data | null,
+      suspense: boolean
+    ): UseFragmentState<Data> => {
+      if (data === null) {
+        return { data: null, fetching: false };
+      }
+
+      const cached = cache.get(request.key);
+      if (!cached) {
+        const newResult = maskFragment<Data>(
+          data,
+          fragment.selectionSet,
+          fragments
+        );
+
+        if (newResult.fulfilled) {
+          return { data: newResult.data, fetching: false };
+        } else if (suspense) {
+          const promise = new Promise(() => {});
```

Comment:
Being able to resolve this Promise might be important for streamed-SSR, I am not sure 😅 the issue here being that we can't keep a stable reference to this cache-entry as we face two challenges

- This could be a series of useFragment, think of a list of todos. This means we'd be able to create a key for the document, however we'd then share this key across all siblings which could be streamed in out of order.
- When we add `data` to the `key` this means that when the previous entry is completed due to the missing part streaming in that we update the key to a new state. This becomes problematic as we never actually resolve the promise, storing the previous reference in a ref isn't possible as React tears down the previous state.
- Leveraging `useId` is also not possible as that isn't stable across Suspense invocations

One solution I can see is deriving a `data` that follows along the document and purposefully omits any `include/skip` and `defer` from the data and uses that to derive the key. This would essentially be similar to creating a `__typename:id` as they key, which could also be relevant here.